### PR TITLE
Convert driver exceptions when starting transactions

### PR DESCRIPTION
**Pull request by lcobucci**
|      Q       |   A
|------------- | -----------
| Type         | bug

#### Summary

The wrapped connection may no longer be valid for long-running processes, and starting a transaction will cause driver exceptions.

DBAL Connection wasn't converting these, leaking internal details of the platform and not providing a consistent API (all other methods would throw a `ConnectionLost` exception for the scenario above).

This ensures we do have the expected behaviour for this edge case.

### CODE REVIEW

**Comment by morozov** [comment]
Let's get it integration-tested. Even though the unit test passes, and the code works as expected on MySQL, we want to be sure that it works the same on other platforms as well or at least know that it doesn't.

File: tests/Functional/TransactionTest.php [hunk 1]
```diff
@@ -65,6 +62,24 @@ private function expectConnectionLoss(callable $scenario): void
...
+    private function killCurrentSession(): void
+    {
+        $databasePlatform = $this->connection->getDatabasePlatform();
+
+        [$currentProcessQuery, $killProcessStatement] = match (true) {
+            $databasePlatform instanceof AbstractMySqlPlatform => ['SELECT CONNECTION_ID()', 'KILL %s'],
+            default => self::markTestSkipped('Unsupported test platform.'),
+        };
+
+        $privilegedConnection = TestUtil::getPrivilegedConnection();
```
Comment (lcobucci) [inline tests/Functional/TransactionTest.php:89]: Using priviledged connection here, assuming that's a requirement for other platforms (maybe oracle or SQL server?)

File: tests/Functional/TransactionTest.php [hunk 2]
```diff
@@ -9,43 +9,40 @@
...
-use function sleep;
+use function sprintf;
 
 use const E_WARNING;
 
 class TransactionTest extends FunctionalTestCase
 {
     public function testCommitFailure(): void
     {
+        $this->markConnectionNotReusable();
```
Comment (lcobucci) [inline tests/Functional/TransactionTest.php]: Marking as not reusable as I didn't want for a failure in one test to leak into the other

File: src/Driver/API/PostgreSQL/ExceptionConverter.php [hunk 3]
```diff
@@ -77,6 +78,13 @@ public function convert(Exception $exception, ?Query $query): DriverException
                 return new ConnectionException($exception, $query);
         }
 
+        if (
+            $exception->getSQLState() === 'HY000'
+            && str_contains($exception->getMessage(), 'server closed the connection')
+        ) {
+            return new ConnectionLost($exception, $query);
+        }
```
Comment (lcobucci) [inline src/Driver/API/PostgreSQL/ExceptionConverter.php]: This might only be applicable for recent PGSQL versions...

File: tests/Functional/TransactionTest.php [hunk 4]
```diff
@@ -21,10 +21,21 @@
 
 class TransactionTest extends FunctionalTestCase
 {
+    public function testBeginTransactionFailure(): void
```
Comment (lcobucci) [inline tests/Functional/TransactionTest.php:24]: And, obviously, `mysqli` behaves differently....

Comment (lcobucci) [reply tests/Functional/TransactionTest.php:24]: 75ced17bc addresses that

File: src/Driver/API/PostgreSQL/ExceptionConverter.php (hunk 3 above)
Comment (lcobucci) [reply src/Driver/API/PostgreSQL/ExceptionConverter.php]: Nah... only PDO has the SQL State set for this error :)

Ooohh I missed working with DBAL :sweat_smile:

File: src/Connection.php [hunk 5]
```diff
@@ -1026,10 +1026,16 @@ public function beginTransaction(): void
 
         ++$this->transactionNestingLevel;
 
-        if ($this->transactionNestingLevel === 1) {
```
Comment (morozov) [inline src/Connection.php:1029]: I wouldn't rearrange this code. Non-nested transactions are more common, and you just wanted to wrap the call to `$connection->beginTransaction()` in a try/catch. Let's not change anything else.

File: tests/Functional/TransactionTest.php [hunk 6]
```diff
@@ -7,45 +7,55 @@
...
-use function sleep;
+use function sprintf;
 
 use const E_WARNING;
 
 class TransactionTest extends FunctionalTestCase
 {
+    public function testBeginTransactionFailure(): void
+    {
+        $this->markConnectionNotReusable();
```
Comment (morozov) [inline tests/Functional/TransactionTest.php]: Would it be fair to also move this call to `killCurrentSession()`? This way, there will be less code duplication and be even more obvious why it's not reusable.

File: tests/Functional/TransactionTest.php [hunk 7]
```diff
@@ -65,6 +75,31 @@ private function expectConnectionLoss(callable $scenario): void
...
+            $databasePlatform instanceof PostgreSQLPlatform => [
+                'SELECT pg_backend_pid()',
+                'SELECT pg_terminate_backend(%s)',
+            ],
+            default => self::markTestSkipped('Unsupported test platform.'),
+        };
+
+        $privilegedConnection = TestUtil::getPrivilegedConnection();
+        $privilegedConnection->executeStatement(
+            sprintf(
```
Comment (morozov) [inline tests/Functional/TransactionTest.php]: Please try passing the PID as a statement parameter.

**Review by morozov — COMMENTED** [review]
The code change and the tests look good. Thanks for replacing `sleep()` with killing the session. Just a few nits inline.

File: src/Connection.php (hunk 5 above)
Comment (lcobucci) [reply src/Connection.php:1029]: Code structure reverted

**Review by morozov — APPROVED** [review]

**Comment by morozov** [comment]
Thanks, @lcobucci!

**Comment by lcobucci** [comment]
@morozov @derrabus would it be alright to release 4.2.3 soon or are you planning to group more items under it?

**Comment by morozov** [comment]
I'm fine with releasing it any time.

**Comment by lcobucci** [comment]
> I'm fine with releasing it any time.

It would be great to have it released, then
